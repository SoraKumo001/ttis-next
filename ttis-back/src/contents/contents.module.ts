import { Module } from '@nestjs/common';
import { ContentsResolver } from './contents.resolver';
import { TypeOrmModule } from '@nestjs/typeorm';
import { Contents } from './contents';
import { ContentsService } from './contents.service';

@Module({
  imports: [TypeOrmModule.forFeature([Contents])],
  providers: [ContentsService,ContentsResolver]
})
export class ContentsModule {}
