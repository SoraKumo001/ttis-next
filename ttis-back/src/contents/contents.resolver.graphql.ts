import { gql } from 'apollo-boost';

const FRAGMENT_CONTENTS = gql`
  fragment FragmentContents on Contents {
    id
    priority
    visible
    page
    title_type
    title
    value_type
    value
    parentId
    createAt
    updateAt
  }
`;

export const QUERY_CONTENTS = gql`
  ${FRAGMENT_CONTENTS}
  query contentsList($id: ID, $level: Int, $visible: Boolean) {
    contentsList(id: $id, level: $level, visible: $visible) {
      ...FragmentContents
    }
  }
`;
export const CREATE_CONTENTS = gql`
  ${FRAGMENT_CONTENTS}
  mutation createContents(
    $parent: ID
    $vector: ContentsVector
    $page: Boolean
    $visible: Boolean
    $title_type: Int
    $title: String
    $value_type: String
    $value: String
  ) {
    createContents(
      parent: $parent
      vector: $vector
      page: $page
      visible: $visible
      title_type: $title_type
      title: $title
      value_type: $value_type
      value: $value
    ) {
      ...FragmentContents
    }
  }
`;
export const UPDATE_CONTENTS = gql`
  ${FRAGMENT_CONTENTS}
  mutation updateContents(
    $id: ID!
    $page: Boolean
    $visible: Boolean
    $title_type: Int
    $title: String
    $parent: ID
    $value_type: String
    $value: String
  ) {
    updateContents(
      id: $id
      page: $page
      visible: $visible
      title_type: $title_type
      title: $title
      parent: $parent
      value_type: $value_type
      value: $value
    ) {
      ...FragmentContents
    }
  }
`;
