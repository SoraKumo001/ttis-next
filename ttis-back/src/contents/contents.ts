import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  CreateDateColumn,
  UpdateDateColumn,
  Tree,
  TreeChildren,
  TreeParent,
} from 'typeorm';
import { ObjectType, Field, ID, Int } from '@nestjs/graphql';

@ObjectType()
@Entity()
@Tree('materialized-path')
export class Contents {
  @Field(() => ID)
  @PrimaryGeneratedColumn('uuid')
  id!: string;
  @Field()
  @Column({ default: 1000 })
  priority!: number;
  @Field({ nullable: true })
  @Column({ nullable: true })
  visible?: boolean;
  @Field()
  @Column({ default: true })
  page!: boolean;
  @Field(() => Int)
  @Column({ default: 1 })
  title_type!: number;
  @Field()
  @Column({ default: 'New' })
  title!: string;
  @Field()
  @Column({ default: 'HTML' })
  value_type!: string;
  @Field()
  @Column({ default: '' })
  value!: string;
  title2?: string;
  @Field({ nullable: true })
  @Column({ nullable: true, default: null })
  parentId?: string;
  @Field(() => [Contents], { nullable: true })
  @TreeChildren()
  children!: Contents[] | null;
  @Field()
  @TreeParent()
  parent?: Contents;
  mpath?: string;
  @Field()
  @CreateDateColumn()
  createAt!: Date;
  @Field()
  @UpdateDateColumn()
  updateAt!: Date;
}

@ObjectType()
export class PageContents {
  @Field(() => ID)
  id!: string;
  @Field(() => [Contents])
  contents!: Contents[];
}
