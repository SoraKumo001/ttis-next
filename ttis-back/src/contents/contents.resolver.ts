import { Resolver, Mutation, ID, Args, Query } from '@nestjs/graphql';
import { Contents } from './contents';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, TreeRepository } from 'typeorm';

@Resolver('Contents')
export class ContentsResolver {
  constructor(
    @InjectRepository(Contents)
    private readonly rep: TreeRepository<Contents>
  ) {
    rep.count().then(count => {
      if (!count) {
        rep.save({ title: 'TOP' });
      }
    });
  }
  @Mutation(_ => ID)
  async create(
    @Args('id', { nullable: true }) parent?: string,
    @Args('vector', { nullable: true }) vector?: number,
    @Args('page', { nullable: true }) page?: boolean,
  ): Promise<string> {
    const { rep } = this;
    if (!parent) {
      parent = (await rep.findOne(undefined, { where: { parentId: null } })).id;
    }
    const contents = await rep.save({
      parent: { id: parent },
      page: !!page,
      title: 'New',
    });
    return contents.id;
  }

  @Query(_ => Contents, { nullable: true })
  async contents(
  ) {
    const { rep } = this;
    return (await rep.findTrees())[0];
  }
}
