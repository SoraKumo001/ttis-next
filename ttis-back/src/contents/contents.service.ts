import { Injectable, OnModuleInit } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Contents } from './contents';
import { ExtendRepository } from '@libs/ExtendRepository';
import { Repository } from 'typeorm';

type VECTOR_TYPE = 'CHILD_FIRST' | 'CHILD_LAST' | 'BEFORE' | 'NEXT';

@Injectable()
export class ContentsService implements OnModuleInit {
  private readonly expRep: ExtendRepository<Contents>;
  constructor(
    @InjectRepository(Contents)
    private readonly rep: Repository<Contents>,
  ) {
    this.expRep = new ExtendRepository(rep.manager.connection, Contents);
  }
  async create(
    id: string,
    vector: VECTOR_TYPE,
    page: boolean,
  ): Promise<Contents|null> {
    const { rep } = this;
    const { parentId, priority } = await this.updateVector(id, vector);
    if (parentId) {
      const contents = await rep.save({
        parent: { id: parentId },
        page: !!page,
        title: 'New',
        priority,
      });
      return contents;
    }
    return null;
  }

  async contents() {
    const { expRep } = this;
    const root = await expRep.findOne({ where: { parent: null } });
    return await expRep.getChildren(root, { order: "priority" });
  }
  async update(contents: Partial<Contents>) {
    const { rep } = this;
    const con = await rep.findOne(contents.id);
    if (con) {
      return rep.save({ ...con, ...contents });
    }
    return null;
  }
  async updateVector(id: string, vector: VECTOR_TYPE) {
    const { rep } = this;
    const baseContents = await rep.findOne({ select: ['id','parentId'], where: { id } });
    let priority: number;
    let parentId: string;
    if (vector === 'BEFORE' || vector === 'NEXT') {
      parentId = baseContents.parentId;
      console.log("parent",parentId);
      if (parentId) {
        const list = await rep.find({
          select: ['id', 'priority'],
          where: { parentId },
          order: { priority: 'ASC' },
        });
        let p = 1;
        for (const contents of list) {
          if (contents.id === id && vector === 'BEFORE') {
            (priority = p), ++p;
          }
          contents.priority = p;
          if (contents.id === id && vector === 'NEXT') {
            ++p;
            priority = p;
          }
          ++p;
        }
        rep.save(list);
      }
    } else if (vector === 'CHILD_FIRST') {
      parentId = baseContents.id;
      if (parentId) {
        const list = await rep.find({
          select: ['id', 'priority'],
          where: { parentId },
          order: { priority: 'ASC' },
        });
        let p = 1;
        for (const contents of list) {
          contents.priority = ++p;
        }
        priority = 1;
        rep.save(list);
      }
    } else {
      parentId = baseContents.id;
      const count = await rep.count({
        parentId: baseContents.id,
      });
      priority = count + 1;
    }
    return { parentId, priority };
  }

  async onModuleInit() {
    const { rep } = this;
    await rep.count().then((count) => {
      if (!count) {
        return rep.save({ title: 'TOP' });
      }
    });
  }
}
