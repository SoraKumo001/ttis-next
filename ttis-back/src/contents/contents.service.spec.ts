import { Test, TestingModule } from '@nestjs/testing';
import { ContentsService } from './contents.service';
import { Contents } from './contents';
import { DBModule2 } from '../db.module';
import { ContentsModule } from './contents.module';
import { ObjectMask } from '@test/testTools';

const masks = {
  id: '',
  parentId: '',
  createAt: '',
  updateAt: '',
};

describe('ContentsService', () => {
  let service: ContentsService;
  let contents: Contents;
  let parentId;
  let module: TestingModule;
  beforeAll(async () => {
    module = await Test.createTestingModule({
      imports: [DBModule2, ContentsModule],
    }).compile();

    service = module.get<ContentsService>(ContentsService);
    await module.init();
  });
  afterAll(async () => {
    await module.close();
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });
  it('List Contents', async () => {
    contents = await service.contents();
    expect(ObjectMask(contents, masks)).toMatchSnapshot();
  });
  it('Create', async () => {
    parentId = (await service.create(contents.id, 'CHILD_FIRST', true)).id;
    expect(!!parentId).toMatchSnapshot();
  });
  it('Create & update', async () => {
    let id: string;
    id = (await service.create(parentId, 'CHILD_LAST', true)).id;
    await service.update({ id, title: 'タイトル1', value: '<div>内容1</div>' });
    id = (await service.create(parentId, 'CHILD_FIRST', true)).id;
    await service.update({ id, title: 'タイトル2', value: '<div>内容2</div>' });
    id = (await service.create(parentId, 'CHILD_LAST', true)).id;
    await service.update({ id, title: 'タイトル3', value: '<div>内容3</div>' });
    id = (await service.create(parentId, 'CHILD_FIRST', true)).id;
    await service.update({ id, title: 'タイトル4', value: '<div>内容4</div>' });

    id = (await service.create(parentId, 'NEXT', true)).id;
    await service.update({
      id,
      title: 'タイトル10',
      value: '<div>内容1</div>',
    });
    id = (await service.create(parentId, 'BEFORE', true)).id;
    await service.update({
      id,
      title: 'タイトル20',
      value: '<div>内容2</div>',
    });
    id = (await service.create(parentId, 'NEXT', true)).id;
    await service.update({
      id,
      title: 'タイトル30',
      value: '<div>内容3</div>',
    });
    id = (await service.create(parentId, 'BEFORE', true)).id;
    await service.update({ id, title: 'タイト40', value: '<div>内容4</div>' });

    expect(!!contents.priority).toMatchSnapshot();
  });
  it('List Contents', async () => {
    contents = await service.contents();
    expect(ObjectMask(contents, masks)).toMatchSnapshot();
  });
});
