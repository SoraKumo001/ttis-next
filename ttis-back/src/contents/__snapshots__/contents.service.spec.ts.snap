// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContentsService Create & update 1`] = `true`;

exports[`ContentsService Create 1`] = `true`;

exports[`ContentsService List Contents 1`] = `
Object {
  "children": Object {},
  "createAt": "",
  "id": "",
  "page": true,
  "parentId": "",
  "priority": 1000,
  "title": "TOP",
  "title_type": 1,
  "updateAt": "",
  "value": "",
  "value_type": "HTML",
  "visible": null,
}
`;

exports[`ContentsService List Contents 2`] = `
Object {
  "children": Object {
    "0": Object {
      "children": Object {},
      "createAt": "",
      "id": "",
      "page": true,
      "parentId": "",
      "priority": 1,
      "title": "タイトル20",
      "title_type": 1,
      "updateAt": "",
      "value": "<div>内容2</div>",
      "value_type": "HTML",
      "visible": null,
    },
    "1": Object {
      "children": Object {},
      "createAt": "",
      "id": "",
      "page": true,
      "parentId": "",
      "priority": 2,
      "title": "タイト40",
      "title_type": 1,
      "updateAt": "",
      "value": "<div>内容4</div>",
      "value_type": "HTML",
      "visible": null,
    },
    "2": Object {
      "children": Object {
        "0": Object {
          "children": Object {},
          "createAt": "",
          "id": "",
          "page": true,
          "parentId": "",
          "priority": 1,
          "title": "タイトル4",
          "title_type": 1,
          "updateAt": "",
          "value": "<div>内容4</div>",
          "value_type": "HTML",
          "visible": null,
        },
        "1": Object {
          "children": Object {},
          "createAt": "",
          "id": "",
          "page": true,
          "parentId": "",
          "priority": 2,
          "title": "タイトル2",
          "title_type": 1,
          "updateAt": "",
          "value": "<div>内容2</div>",
          "value_type": "HTML",
          "visible": null,
        },
        "2": Object {
          "children": Object {},
          "createAt": "",
          "id": "",
          "page": true,
          "parentId": "",
          "priority": 3,
          "title": "タイトル1",
          "title_type": 1,
          "updateAt": "",
          "value": "<div>内容1</div>",
          "value_type": "HTML",
          "visible": null,
        },
        "3": Object {
          "children": Object {},
          "createAt": "",
          "id": "",
          "page": true,
          "parentId": "",
          "priority": 4,
          "title": "タイトル3",
          "title_type": 1,
          "updateAt": "",
          "value": "<div>内容3</div>",
          "value_type": "HTML",
          "visible": null,
        },
      },
      "createAt": "",
      "id": "",
      "page": true,
      "parentId": "",
      "priority": 3,
      "title": "New",
      "title_type": 1,
      "updateAt": "",
      "value": "",
      "value_type": "HTML",
      "visible": null,
    },
    "3": Object {
      "children": Object {},
      "createAt": "",
      "id": "",
      "page": true,
      "parentId": "",
      "priority": 4,
      "title": "タイトル30",
      "title_type": 1,
      "updateAt": "",
      "value": "<div>内容3</div>",
      "value_type": "HTML",
      "visible": null,
    },
    "4": Object {
      "children": Object {},
      "createAt": "",
      "id": "",
      "page": true,
      "parentId": "",
      "priority": 5,
      "title": "タイトル10",
      "title_type": 1,
      "updateAt": "",
      "value": "<div>内容1</div>",
      "value_type": "HTML",
      "visible": null,
    },
  },
  "createAt": "",
  "id": "",
  "page": true,
  "parentId": "",
  "priority": 1000,
  "title": "TOP",
  "title_type": 1,
  "updateAt": "",
  "value": "",
  "value_type": "HTML",
  "visible": null,
}
`;
